discord/message.py

```diff
@@ -843,6 +847,51 @@ def __init__(self, data: RoleSubscriptionDataPayload) -> None:
         self.is_renewal: bool = data['is_renewal']
 
 
+class GuildProductPurchase:
+    """Represents a message's guild product that the user has purchased.
+
+    .. versionadded:: 2.5
+
+    Attributes
+    -----------
+    listing_id: :class:`int`
+        The ID of the listing that the user has purchased.
+    product_name: :class:`str`
+        The name of the product that the user has purchased.
+    """
+
+    def __init__(self, data: GuildProductPurchasePayload) -> None:
+        self.listing_id: int = int(data['listing_id'])
+        self.product_name: str = data['product_name']
+
+    def __hash__(self) -> int:
+        return self.listing_id >> 22
```

Comment:
`__hash__` with no `__eq__` is weird, you should define both if you're going to do that.